### PR TITLE
#1869 Fix null reference errors when training from scratch with CUDA

### PR DESCRIPTION
**Pull request by robyngraf**
Calling ggml_compute_forward when node->src0 was null was causing train-text-from-scratch.exe to terminate unexpectedly if it was compiled with CUDA on.

### CODE REVIEW

**Comment by KerfuffleV2** [comment]
This does seem to fix the problem for me. I can't say much about the results (and I didn't let it run to completion) but `master` just segfaults immediately while this PR seems to run without a problem.

It doesn't seem like it really uses the GPU but I'm assuming that was the case regardless of this issue.

**Comment by ggerganov** [comment]
Which op has `src0 == NULL`? This might indicate an issue in the training code and should be fixed there.
With the proposed change, we might be potentially "hiding" it, which is not desired

**Comment by JohannesGaessler** [comment]
This is the GDB output that I get:

```
(gdb) l
2392        g_scratch_buffer = nullptr;
2393    }
2394
2395    bool ggml_cuda_compute_forward(struct ggml_compute_params * params, struct ggml_tensor * tensor){
2396        ggml_cuda_func_t func;
2397        const bool any_on_device = tensor->backend == GGML_BACKEND_GPU
2398            || tensor->src0->backend == GGML_BACKEND_GPU || tensor->src0->backend == GGML_BACKEND_GPU_SPLIT
2399            || (tensor->src1 != nullptr && tensor->src1->backend == GGML_BACKEND_GPU);
2400
2401        switch (tensor->op) {
(gdb) p *tensor
$6 = {type = GGML_TYPE_F32, backend = GGML_BACKEND_CPU, n_dims = 2, ne = {256, 32000, 1, 1}, nb = {4, 1024, 32768000, 32768000}, 
  op = GGML_OP_NONE, is_param = true, grad = 0x7ff93af297d0, src0 = 0x0, src1 = 0x0, opt = {0x0, 0x0, 0x0, 0x0}, n_tasks = 1, perf_runs = 0, 
  perf_cycles = 0, perf_time_us = 0, data = 0x7fff01fff110, name = "tok_embeddings.weight\000\000\000\000\000\000\000\000\000\000", extra = 0x0, 
  padding = "\000\000\000"}
(gdb) bt
#0  0x00005555555fd361 in ggml_cuda_compute_forward (params=0x7ffffffcce40, tensor=0x7fff01fff030)
    at /home/johannesg/Projects/llama.cpp/ggml-cuda.cu:2397
#1  0x00005555555ab59f in ggml_compute_forward (params=0x7ffffffcce40, tensor=0x7fff01fff030) at ggml.c:14308
#2  0x00005555555b0309 in ggml_graph_compute (ctx=0x5555556fab08 <g_state+104>, cgraph=0x7ff911a00510) at ggml.c:16009
#3  0x000055555556c2a5 in main (argc=5, argv=0x7fffffffd768) at examples/train-text-from-scratch/train-text-from-scratch.cpp:3258
(gdb) 
```

The segfault occurs when `ggml_cuda_compute_forward` tries to determine whether or not `embeddings` should be run on the GPU. `embeddings` has `GGML_OP_NONE` so I'm not sure why `ggml_compute_forward` is being called for it. If `ggml_compute_forward` should be called for this tensor then I think the correct fix would be to modify `ggml_cuda_compute_forward`.

**Comment by dranger003** [comment]
> This is the GDB output that I get:
> 
> ```
> (gdb) l
> 2392        g_scratch_buffer = nullptr;
> 2393    }
> 2394
> 2395    bool ggml_cuda_compute_forward(struct ggml_compute_params * params, struct ggml_tensor * tensor){
> 2396        ggml_cuda_func_t func;
> 2397        const bool any_on_device = tensor->backend == GGML_BACKEND_GPU
> 2398            || tensor->src0->backend == GGML_BACKEND_GPU || tensor->src0->backend == GGML_BACKEND_GPU_SPLIT
> 2399            || (tensor->src1 != nullptr && tensor->src1->backend == GGML_BACKEND_GPU);
> 2400
> 2401        switch (tensor->op) {
> (gdb) p *tensor
> $6 = {type = GGML_TYPE_F32, backend = GGML_BACKEND_CPU, n_dims = 2, ne = {256, 32000, 1, 1}, nb = {4, 1024, 32768000, 32768000}, 
>   op = GGML_OP_NONE, is_param = true, grad = 0x7ff93af297d0, src0 = 0x0, src1 = 0x0, opt = {0x0, 0x0, 0x0, 0x0}, n_tasks = 1, perf_runs = 0, 
>   perf_cycles = 0, perf_time_us = 0, data = 0x7fff01fff110, name = "tok_embeddings.weight\000\000\000\000\000\000\000\000\000\000", extra = 0x0, 
>   padding = "\000\000\000"}
> (gdb) bt
> #0  0x00005555555fd361 in ggml_cuda_compute_forward (params=0x7ffffffcce40, tensor=0x7fff01fff030)
>     at /home/johannesg/Projects/llama.cpp/ggml-cuda.cu:2397
> #1  0x00005555555ab59f in ggml_compute_forward (params=0x7ffffffcce40, tensor=0x7fff01fff030) at ggml.c:14308
> #2  0x00005555555b0309 in ggml_graph_compute (ctx=0x5555556fab08 <g_state+104>, cgraph=0x7ff911a00510) at ggml.c:16009
> #3  0x000055555556c2a5 in main (argc=5, argv=0x7fffffffd768) at examples/train-text-from-scratch/train-text-from-scratch.cpp:3258
> (gdb) 
> ```
> 
> The segfault occurs when `ggml_cuda_compute_forward` tries to determine whether or not `embeddings` should be run on the GPU. `embeddings` has `GGML_OP_NONE` so I'm not sure why `ggml_compute_forward` is being called for it. If `ggml_compute_forward` should be called for this tensor then I think the correct fix would be to modify `ggml_cuda_compute_forward`.

Fixing this one isn't enough as then it crashes in ggml.c (see below). If both are fixed, then the training runs.

https://github.com/ggerganov/llama.cpp/blob/d411968e990c37f51328849c96a743dd78f3c3dd/ggml-cuda.cu#L2555
https://github.com/ggerganov/llama.cpp/blob/d411968e990c37f51328849c96a743dd78f3c3dd/ggml.c#L14318

https://github.com/ggerganov/llama.cpp/pull/1652#issuecomment-1595570310

**Comment by xaedes** [comment]
Will take a closer look onto that later this weekend.

At first glance it seems the issue is accessing src0 without checking that src0 is NULL. GGML_OP_NONE tensors have no source tensors, which is correct.

In forward-only this issue does not appear because those tensors are marked as leafs by `ggml_visit_parents`, so they are not processed by `ggml_compute_forward`.

But when they have a gradient they are not marked as leaf because the condition for leafs is `(node->op == GGML_OP_NONE && node->grad == NULL)`.

This condition make some sense, because when they have a grad, they are required to be nodes, because only nodes are processed by `ggml_build_backward`.
On the other hand, they don't really have to be processed by `ggml_build_backward` when they have no source, because adding something to the source gradients is all that `ggml_build_backward` is doing.

So maybe a change to the leaf condition could be useful.

But still.. It is very easy to imagine that we one day have a ggml operation without inputs, for example to generate random values, or to generate a tensor filled with a sequence of integers or similar. Then the access to src0 will segfault again.

So even when changing the condition for leafs it makes sense to check if src0 != NULL before accessing it.

But the check should be immediatly before accessing it. Basically line 14318 should have a check like the line after it:
https://github.com/ggerganov/llama.cpp/blob/051e1b0e6a6e3aee7d989b47760980e6fda5861c/ggml.c#L14318-L14319

So that it looks like this:
```cpp
 GGML_ASSERT(tensor->src0 == NULL || tensor->src0->backend == GGML_BACKEND_CPU); 
 GGML_ASSERT(tensor->src1 == NULL || tensor->src1->backend == GGML_BACKEND_CPU); 
```

Similar in ggml-cuda.cu:
https://github.com/ggerganov/llama.cpp/blob/051e1b0e6a6e3aee7d989b47760980e6fda5861c/ggml-cuda.cu#L2558-L2560

Should look like this:

```cpp
    const bool any_on_device = tensor->backend == GGML_BACKEND_GPU
        || (tensor->src0 != nullptr && (tensor->src0->backend == GGML_BACKEND_GPU || tensor->src0->backend == GGML_BACKEND_GPU_SPLIT))
        || (tensor->src1 != nullptr && tensor->src1->backend == GGML_BACKEND_GPU);
```

These are my first thoughts on this, will take a closer look into that later.

**Comment by howard0su** [comment]
I would propose the following fix:

@@ -14330,6 +14330,10 @@ static void ggml_compute_forward_cross_entropy_loss_back(
 static void ggml_compute_forward(struct ggml_compute_params * params, struct ggml_tensor * tensor) {
     GGML_ASSERT(params);

+    if (tensor->op == GGML_OP_NONE)
+        return;
+
+

**Review by xaedes — CHANGES_REQUESTED** [review]
Sorry, for the long wait.

I now checked whether changing the leaf condition is viable.
We should not change it, since the optimization functions only optimizes tensors that are contained in cgraph->nodes.
Additionally changing the condition would also change the semantic meaning of "leaf" which is "a leaf node, not part of the gradient graph (e.g. a constant)".

So the GGML_OP_NONE tensors with gradients must indeed be nodes, not leafs.

This means that compute_forward must expect to be called on GGML_OP_NONE tensors.
This is no issue, as the switch case will make it so that nothing is done for these tensors.

Only the unchecked access to src0 is problematic. 
The implicit assumption that `src0 != NULL` is not true.

We should just add a test whether src0 is NULL before accessing it. 

@robyngraf Can you change the pull request, as following?

Revert the conditional execution of ggml_compute_forward. Just `ggml_compute_forward(&params, node);` instead of  `if (node->src0) { ggml_compute_forward(&params, node); }`.

Add src0 test to the condition in `ggml.c` https://github.com/ggerganov/llama.cpp/blob/4e3b9f2f9c9f097b1771888dcf614e291dab9c67/ggml.c#L14338-L14339, so that it looks like this:

```cpp
GGML_ASSERT(tensor->src0 == NULL || tensor->src0->backend == GGML_BACKEND_CPU);
GGML_ASSERT(tensor->src1 == NULL || tensor->src1->backend == GGML_BACKEND_CPU);
```

Add src0 test to the condition in `ggml-cuda.cu` https://github.com/ggerganov/llama.cpp/blob/4e3b9f2f9c9f097b1771888dcf614e291dab9c67/ggml-cuda.cu#L2544-L2546, so that it looks like this:

```cpp
const bool any_on_device = tensor->backend == GGML_BACKEND_GPU
    || (tensor->src0 != nullptr && (tensor->src0->backend == GGML_BACKEND_GPU || tensor->src0->backend == GGML_BACKEND_GPU_SPLIT))
    || (tensor->src1 != nullptr && tensor->src1->backend == GGML_BACKEND_GPU);
```

**Review by ggerganov — APPROVED** [review]
I made the changes proposed by @xaedes 

Can't test on CUDA right now. If this works - merge it

**Comment by howard0su** [comment]
verified on P100.

**Review by xaedes — APPROVED** [review]
